pkg/service/repair/plan.go

```diff
@@ -13,124 +13,138 @@ import (
+			tables    []tablePlan
+			rangesCnt int
+		)
 		for _, rep := range ring.ReplicaTokens {
-			rtr := scyllaclient.ReplicaTokenRanges{
-				ReplicaSet: filteredReplicaSet(rep.ReplicaSet, filtered, target.Host),
-				Ranges:     rep.Ranges,
+			for _, t := range u.Tables {
+				for _, h := range rep.ReplicaSet {
+					k := hkt{
+						Host:     h,
+						Keyspace: u.Keyspace,
+						Table:    t,
+					}
+					hostRanges[k] += len(rep.Ranges)
+				}
 			}
-
-			// Don't add keyspace with some ranges not replicated in filtered hosts,
-			// unless it's a single node cluster.
-			if len(rtr.ReplicaSet) <= 1 && len(status) > 1 {
-				skip = true
-				break
-			}
-
-			for _, r := range rtr.Ranges {
-				kp.TokenRepIdx[r] = len(kp.Replicas)
-			}
-			kp.Replicas = append(kp.Replicas, rtr)
+			sets = append(sets, rep.ReplicaSet)
+			allHosts.Add(rep.ReplicaSet...)
+			rangesCnt += len(rep.Ranges)
 		}
 
-		if skip {
-			p.SkippedKeyspaces = append(p.SkippedKeyspaces, u.Keyspace)
-			continue
+		if p := maxParallel(dcMap, sets); maxP < p {
+			maxP = p
 		}
 
-		// Fill tables
 		for _, t := range u.Tables {
-			kp.Tables = append(kp.Tables, tablePlan{
-				Table:           t,
-				MarkedRanges:    make(map[scyllaclient.TokenRange]struct{}),
-				MarkedInReplica: make([]int, len(kp.Replicas)),
+			tables = append(tables, tablePlan{
+				Table:         t,
+				Size:          -1, // Filled with fillSize
+				ReplicaSetCnt: len(sets),
+				RangesCnt:     rangesCnt,
+				Optimize:      false, // Filled with fillOptimize
 			})
 		}
-		p.Keyspaces = append(p.Keyspaces, kp)
+		keyspaces = append(keyspaces, keyspacePlan{
+			Keyspace: u.Keyspace,
+			Size:     -1, // Filled with fillSize
+			Tables:   tables,
+		})
 	}
 
-	if len(p.Keyspaces) == 0 {
+	if len(keyspaces) == 0 {
 		return nil, ErrEmptyRepair
 	}
-	if err := p.FillSize(ctx, client, target.SmallTableThreshold); err != nil {
-		return nil, errors.Wrap(err, "calculate tables size")
-	}
-	return p, nil
-}
+	hosts := allHosts.List()
 
-// UpdateIdx sets keyspace and table idx to the next not repaired table.
-// Returns false if there are no more tables to repair.
-func (p *plan) UpdateIdx() bool {
-	ksIdx := p.Idx
-	tabIdx := p.Keyspaces[ksIdx].Idx
-
-	for ; ksIdx < len(p.Keyspaces); ksIdx++ {
-		kp := p.Keyspaces[ksIdx]
-		for ; tabIdx < len(kp.Tables); tabIdx++ {
-			// Always wait for current table to be fully repaired before moving to the next one
-			if !kp.IsTableRepaired(tabIdx) {
-				p.Idx = ksIdx
-				p.Keyspaces[ksIdx].Idx = tabIdx
-				return true
+	var hkts []hkt
+	for _, ksp := range keyspaces {
+		for _, tp := range ksp.Tables {
+			for _, h := range hosts {
+				hkts = append(hkts, hkt{Host: h, Keyspace: ksp.Keyspace, Table: tp.Table})
 			}
 		}
-		tabIdx = 0
 	}
+	hostSize, err := perHostTableSize(ctx, client, hkts)
+	if err != nil {
+		return nil, errors.Wrap(err, "calculate tables size")
+	}
+	fillSize(hostSize, keyspaces)
+	fillOptimize(keyspaces, target.SmallTableThreshold)
```

Comment:
What do you think about moving `fillOptimize` under `keyspaces` authority ? 
...and to call `keyspaces.fillOptimize` ? (as the comment on fillSize)